DOC/DEV: correct editable install `spin` guidance

The current guidance seems at least partly false.

https://github.com/scipy/scipy/blob/abb598d7e5634cddb903a0ad7b42f8b57c895719/doc/source/building/index.rst?plain=1#L387-L419

For example:

```console
scipy on 🎋 main is 📦 v1.17.0.dev0 via 🐍 v3.12.11 via 🧚 v0.50.0 (default) 
❯ spin install
$ pip install -v --config-settings=editable-verbose=true --no-build-isolation --editable .
...

scipy on 🎋 main is 📦 v1.17.0.dev0 via 🐍 v3.12.11 via 🧚 v0.50.0 (default)
❯ spin test -s cluster
Invoking `build` prior to running tests:
...
Editable install of same source detected; skipping build
Editable install of same source directory detected; not setting PYTHONPATH
...
=========================================================================================== test session starts ============================================================================================
platform darwin -- Python 3.12.11, pytest-8.4.1, pluggy-1.6.0
rootdir: /Users/lucascolley/ghq/github.com/lucascolley/pixi-dev-scipystack/scipy/scipy
configfile: pytest.ini
plugins: xdist-3.8.0, timeout-2.4.0, cov-6.2.1, hypothesis-6.135.32
collected 217 items                                                                                                                                                                                        

scipy/cluster/tests/test_disjoint_set.py .....................                                                                                                                                       [  9%]
scipy/cluster/tests/test_hierarchy.py ......................................................................................................................................................ssssss.. [ 82%]
.....                                                                                                                                                                                                [ 84%]
scipy/cluster/tests/test_vq.py .................................                                                                                                                                     [100%]

====================================================================================== 211 passed, 6 skipped in 6.59s ======================================================================================
```

---

Am I right in thinking that with `spin`, pretty much everything should 'just work' when performing an editable installation with `spin install` instead of `spin build`? What about IDE support?

cc @rgommers @stefanv. Maybe you are also using this @mdhaber?

Following up on https://github.com/scipy/scipy/issues/23383#issuecomment-3117946047.

Like before, the editable install seems to be working pretty well. Most of my problems seem to be related to Windows rather than editable install.

`install`, `python`, `ipython`, `lint`, `refguide-check`, `spin smoke-tutorials` all worked without trouble.

`spin smoke-tutorials` has one failure for me, though - a Windows thing; something is looking for `/usr/share/dict/words`.

<details>

```
________________________________________________ [doctest] csgraph.rst ________________________________________________
031     /usr/share/dict
032     /var/lib/dict
033
034 Another easy source for words are the Scrabble word lists available at various
035 sites around the internet (search with your favorite search engine). We'll
036 first create this list. The system word lists consist of a file with one
037 word per line. The following should be modified to use the particular word
038 list you have available::
039
040     >>> with open('/usr/share/dict/words') as f:
UNEXPECTED EXCEPTION: FileNotFoundError(2, 'No such file or directory')
Traceback (most recent call last):
  File "C:\Users\matth\miniforge3\envs\scipy-dev\Lib\doctest.py", line 1395, in __run
    exec(compile(example.source, filename, "single",
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                 compileflags, True), test.globs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<doctest csgraph.rst[0]>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '/usr/share/dict/words'
C:\Users\matth\Desktop\scipy\doc\source\tutorial\csgraph.rst:40: UnexpectedException
=============================================== short test summary info ===============================================
FAILED doc/source/tutorial/csgraph.rst::csgraph.rst
====================================== 1 failed, 26 passed, 5 xfailed in 36.99s =======================================
```

</details>

I also get this message for basically all commands.
<details>

```
C:\Users\matth\Desktop\scipy>spin python
Invoking `build` prior to invoking Python:
C:\Users\matth\Desktop\scipy\.spin/cmds.py:138: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] The system cannot find the file specified
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  n_cores = cpu_count(only_physical_cores=True)
  File "C:\Users\matth\Desktop\scipy\.spin/cmds.py", line 956, in _count_physical_cores
    cpu_info = subprocess.run(
        "wmic CPU Get NumberOfCores /Format:csv".split(),
        capture_output=True,
        text=True,
    )
  File "C:\Users\matth\miniforge3\envs\scipy-dev\Lib\subprocess.py", line 554, in run
    with Popen(*popenargs, **kwargs) as process:
         ~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\matth\miniforge3\envs\scipy-dev\Lib\subprocess.py", line 1039, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                        pass_fds, cwd, env,
                        ^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
                        gid, gids, uid, umask,
                        ^^^^^^^^^^^^^^^^^^^^^^
                        start_new_session, process_group)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\matth\miniforge3\envs\scipy-dev\Lib\subprocess.py", line 1551, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
                             # no special security
                             ^^^^^^^^^^^^^^^^^^^^^
    ...<4 lines>...
                             cwd,
                             ^^^^
                             startupinfo)
                             ^^^^^^^^^^^^
Editable install of same source detected; skipping build
Editable install of same source directory detected; not setting PYTHONPATH
$ 'C:\Users\matth\miniforge3\envs\scipy-dev\python.exe' -P
Python 3.13.3 | packaged by conda-forge | (main, Apr 14 2025, 20:31:24) [MSC v.1943 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
```

The part about `Editable install` is red - a little alarming for something that I think is as-expected.

</details>

`spin bench` doesn't work out of the box, but I think this is a `conda` `asv` thing; I've always had bad luck with that.
```
$ asv run --dry-run --show-stderr --python=same
Fatal error in launcher: Unable to create process using '"D:\bld\asv_1725737853619\_h_env\python.exe"  "C:\Users\matth\miniforge3\envs\scipy-dev\Scripts\asv.exe" run --dry-run --show-stderr --python=same': The system cannot find the file specified
```
When I install `asv` with `pip`, benchmarks run.

`spin docs` doesn't work for me - but initially it's because `make` doesn't work. When I install `make`:
```
$ export SPHINXOPTS=-W -D plot_gallery=0 -j auto
$ cd doc
$ make html
The system cannot find the path specified.
The system cannot find the path specified.
# @echo sdist build with unlabeled sources
process_begin: CreateProcess(NULL, # @echo sdist build with unlabeled sources, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [Makefile:62: version-check] Error 2
```
Not surprising; I've never successfully built docs on Windows.
`spin shell` doesn't work... also a Windows thing (`'sh' executable not found. Exiting.`)
`spin gdb` and `spin lldb` don't work because I don't have them. Don't need them - editable installs can debug in an IDE.
`spin mypy` doesn't work as Ralf mentioned: `Error: Mypy does not work (well) for editable installs`.
After installing `scipy-doctest` (which should probably be in `environment.yml`?), `spin smoke-docs` fails during collection.
<details>

```
C:\Users\matth\Desktop\scipy>spin smoke-docs
Invoking `build` prior to running tests:
C:\Users\matth\Desktop\scipy\.spin/cmds.py:138: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] The system cannot find the file specified
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  n_cores = cpu_count(only_physical_cores=True)
  File "C:\Users\matth\Desktop\scipy\.spin/cmds.py", line 956, in _count_physical_cores
    cpu_info = subprocess.run(
        "wmic CPU Get NumberOfCores /Format:csv".split(),
        capture_output=True,
        text=True,
    )
  File "C:\Users\matth\miniforge3\envs\scipy-dev\Lib\subprocess.py", line 554, in run
    with Popen(*popenargs, **kwargs) as process:
         ~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\matth\miniforge3\envs\scipy-dev\Lib\subprocess.py", line 1039, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                        pass_fds, cwd, env,
                        ^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
                        gid, gids, uid, umask,
                        ^^^^^^^^^^^^^^^^^^^^^^
                        start_new_session, process_group)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\matth\miniforge3\envs\scipy-dev\Lib\subprocess.py", line 1551, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
                             # no special security
                             ^^^^^^^^^^^^^^^^^^^^^
    ...<4 lines>...
                             cwd,
                             ^^^^
                             startupinfo)
                             ^^^^^^^^^^^^
Editable install of same source detected; skipping build
Editable install of same source directory detected; not setting PYTHONPATH
$ 'C:\Users\matth\miniforge3\envs\scipy-dev\python.exe' -P -c 'import scipy'
meson-python: building scipy: meson compile
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: C:\Users\matth\miniforge3\envs\scipy-dev\Library\bin\ninja.EXE
[2/338] Generating subprojects/highs/src/HConfig.h with a custom command
$ 'C:\Users\matth\miniforge3\envs\scipy-dev\python.exe' -P -m pytest -m 'not slow' scipy --doctest-modules --doctest-only-doctests=true --doctest-collect=api
meson-python: building scipy: meson compile
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: C:\Users\matth\miniforge3\envs\scipy-dev\Library\bin\ninja.EXE
[2/338] Generating subprojects/highs/src/HConfig.h with a custom command
================================================= test session starts =================================================
platform win32 -- Python 3.13.3, pytest-8.3.5, pluggy-1.6.0
rootdir: C:\Users\matth\Desktop\scipy
configfile: pytest.ini
plugins: anyio-4.9.0, hypothesis-6.131.20, cov-6.1.1, skip-slow-0.0.5, timeout-2.4.0, xdist-3.6.1, scipy_doctest-1.8.0
collected 4364 items / 13 errors

======================================================= ERRORS ========================================================
___________________________________ ERROR collecting scipy/_build_utils/tempita.py ____________________________________
import file mismatch:
imported module 'tempita' has this __file__ attribute:
  C:\Users\matth\Desktop\scipy\scipy\_build_utils\tempita
which is not the same as the test file we want to collect:
  C:\Users\matth\Desktop\scipy\scipy\_build_utils\tempita.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
________________________________ ERROR collecting scipy/interpolate/_interpnd_info.py _________________________________
ImportError while importing test module 'C:\Users\matth\Desktop\scipy\scipy\interpolate\_interpnd_info.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
..\..\miniforge3\envs\scipy-dev\Lib\importlib\__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'scipy.interpolate._interpnd_info'
______________________________ ERROR collecting scipy/interpolate/_rbfinterp_pythran.py _______________________________
import file mismatch:
imported module 'scipy.interpolate._rbfinterp_pythran' has this __file__ attribute:
  C:\Users\matth\Desktop\scipy\build\cp313\scipy\interpolate\_rbfinterp_pythran.cp313-win_amd64.pyd
which is not the same as the test file we want to collect:
  C:\Users\matth\Desktop\scipy\scipy\interpolate\_rbfinterp_pythran.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
____________________________ ERROR collecting scipy/linalg/_cython_signature_generator.py _____________________________
ImportError while importing test module 'C:\Users\matth\Desktop\scipy\scipy\linalg\_cython_signature_generator.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
..\..\miniforge3\envs\scipy-dev\Lib\importlib\__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'scipy.linalg._cython_signature_generator'
___________________________________ ERROR collecting scipy/linalg/_generate_pyx.py ____________________________________
ImportError while importing test module 'C:\Users\matth\Desktop\scipy\scipy\linalg\_generate_pyx.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
..\..\miniforge3\envs\scipy-dev\Lib\importlib\__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'scipy.linalg._generate_pyx'
__________________________________ ERROR collecting scipy/linalg/_linalg_pythran.py ___________________________________
import file mismatch:
imported module 'scipy.linalg._linalg_pythran' has this __file__ attribute:
  C:\Users\matth\Desktop\scipy\build\cp313\scipy\linalg\_linalg_pythran.cp313-win_amd64.pyd
which is not the same as the test file we want to collect:
  C:\Users\matth\Desktop\scipy\scipy\linalg\_linalg_pythran.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
________________________________ ERROR collecting scipy/linalg/_matfuncs_sqrtm_triu.py ________________________________
import file mismatch:
imported module 'scipy.linalg._matfuncs_sqrtm_triu' has this __file__ attribute:
  C:\Users\matth\Desktop\scipy\build\cp313\scipy\linalg\_matfuncs_sqrtm_triu.cp313-win_amd64.pyd
which is not the same as the test file we want to collect:
  C:\Users\matth\Desktop\scipy\scipy\linalg\_matfuncs_sqrtm_triu.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
_________________________ ERROR collecting scipy/ndimage/utils/generate_label_testvectors.py __________________________
scipy\ndimage\utils\generate_label_testvectors.py:44: in <module>
    generate_test_vecs("label_inputs.txt", "label_strels.txt", "label_results.txt")
        __builtins__ = <builtins>
        __cached__ = 'C:\\Users\\matth\\Desktop\\scipy\\scipy\\ndimage\\utils\\__pycache__\\generate_label_testvectors.cpython-313.pyc'
        __doc__    = None
        __file__   = 'C:\\Users\\matth\\Desktop\\scipy\\scipy\\ndimage\\utils\\generate_label_testvectors.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x000001A18D1DFB30>
        __name__   = 'generate_label_testvectors'
        __package__ = ''
        __spec__   = ModuleSpec(name='generate_label_testvectors', loader=<_frozen_importlib_external.SourceFileLoader object at 0x000001A18D1DFB30>, origin='C:\\Users\\matth\\Desktop\\scipy\\scipy\\ndimage\\utils\\generate_label_testvectors.py')
        __warningregistry__ = {'version': 4283}
        generate_test_vecs = <function generate_test_vecs at 0x000001A18D15F740>
        label      = <function label at 0x000001A18A997420>
        np         = <module 'numpy' from 'C:\\Users\\matth\\miniforge3\\envs\\scipy-dev\\Lib\\site-packages\\numpy\\__init__.py'>
scipy\ndimage\utils\generate_label_testvectors.py:34: in generate_test_vecs
    strels = [np.fromstring(s, dtype=int).reshape((3, 3))
E   DeprecationWarning: The binary mode of fromstring is deprecated, as it behaves surprisingly on unicode inputs. Use frombuffer instead
        bitimage   = <function generate_test_vecs.<locals>.bitimage at 0x000001A18D15F600>
        data       = [array([[1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
      ...,
       [False, False, False,  True, False, False, False],
       [ True, False,  True,  True,  True, False,  True]])]
        infile     = 'label_inputs.txt'
        resultfile = 'label_results.txt'
        strelfile  = 'label_strels.txt'
        strels     = [array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]]), array([[0., 0., 0.],
       [0., 0., 0.],
       [0....alse, False,  True]]), array([[False, False, False],
       [ True,  True,  True],
       [False, False, False]]), ...]
__________________________________ ERROR collecting scipy/optimize/_group_columns.py __________________________________
import file mismatch:
imported module 'scipy.optimize._group_columns' has this __file__ attribute:
  C:\Users\matth\Desktop\scipy\build\cp313\scipy\optimize\_group_columns.cp313-win_amd64.pyd
which is not the same as the test file we want to collect:
  C:\Users\matth\Desktop\scipy\scipy\optimize\_group_columns.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
_________________________________ ERROR collecting scipy/signal/_max_len_seq_inner.py _________________________________
import file mismatch:
imported module 'scipy.signal._max_len_seq_inner' has this __file__ attribute:
  C:\Users\matth\Desktop\scipy\build\cp313\scipy\signal\_max_len_seq_inner.cp313-win_amd64.pyd
which is not the same as the test file we want to collect:
  C:\Users\matth\Desktop\scipy\scipy\signal\_max_len_seq_inner.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
_______________________________ ERROR collecting scipy/sparse/_generate_sparsetools.py ________________________________
ImportError while importing test module 'C:\Users\matth\Desktop\scipy\scipy\sparse\_generate_sparsetools.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
..\..\miniforge3\envs\scipy-dev\Lib\importlib\__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'scipy.sparse._generate_sparsetools'
___________________________________ ERROR collecting scipy/special/_generate_pyx.py ___________________________________
ImportError while importing test module 'C:\Users\matth\Desktop\scipy\scipy\special\_generate_pyx.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
..\..\miniforge3\envs\scipy-dev\Lib\importlib\__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'scipy.special._generate_pyx'
___________________________________ ERROR collecting scipy/stats/_stats_pythran.py ____________________________________
import file mismatch:
imported module 'scipy.stats._stats_pythran' has this __file__ attribute:
  C:\Users\matth\Desktop\scipy\build\cp313\scipy\stats\_stats_pythran.cp313-win_amd64.pyd
which is not the same as the test file we want to collect:
  C:\Users\matth\Desktop\scipy\scipy\stats\_stats_pythran.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
=============================================== short test summary info ===============================================
ERROR scipy/_build_utils/tempita.py
ERROR scipy/interpolate/_interpnd_info.py
ERROR scipy/interpolate/_rbfinterp_pythran.py
ERROR scipy/linalg/_cython_signature_generator.py
ERROR scipy/linalg/_generate_pyx.py
ERROR scipy/linalg/_linalg_pythran.py
ERROR scipy/linalg/_matfuncs_sqrtm_triu.py
ERROR scipy/ndimage/utils/generate_label_testvectors.py - DeprecationWarning: The binary mode of fromstring is deprecated, as it behaves surprisingly on unicode inputs. Use ...
ERROR scipy/optimize/_group_columns.py
ERROR scipy/signal/_max_len_seq_inner.py
ERROR scipy/sparse/_generate_sparsetools.py
ERROR scipy/special/_generate_pyx.py
ERROR scipy/stats/_stats_pythran.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 13 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================= 13 errors in 6.92s ==================================================
```

</details>

--- a/doc/source/building/index.rst
+++ b/doc/source/building/index.rst
@@ -395,7 +395,7 @@ interface is self-documenting, so please see ``spin --help`` and
 
     Editable installs are supported via ``spin install``.
     When making changes to SciPy code, including to compiled code, there is no
-    need to manually rebuild or reinstall. When you run ``git clean -xdf``,
+    need to manually rebuild or reinstall. However, should you need to run ``git clean -xdf``,
     which removes the built extension modules, remember to also uninstall SciPy
     with ``pip uninstall scipy``.
 

--- a/doc/source/building/index.rst
+++ b/doc/source/building/index.rst
@@ -386,30 +386,14 @@ interface is self-documenting, so please see ``spin --help`` and
 
 .. admonition:: IDE support & editable installs
 
-    While the ``spin`` interface is our recommended way of working on SciPy,
+    While the ``spin build`` interface is our recommended way of working on SciPy,
     it has one limitation: because of the custom install location, SciPy
-    installed using ``spin`` will not be recognized automatically within an
+    will not be recognized automatically within an
     IDE (e.g., for running a script via a "run" button, or setting breakpoints
     visually). This will work better with an *in-place build* (or "editable
     install").
 
-    Editable installs are supported. It is important to understand that **you
-    may use either an editable install or spin in a given repository clone,
-    but not both**. If you use editable installs, you have to use ``pytest``
-    and other development tools directly instead of using ``spin``.
-
-    To use an editable install, ensure you start from a clean repository (run
-    ``git clean -xdf`` if you've built with ``spin`` before) and have all
-    dependencies set up correctly as described higher up on this page. Then
-    do::
-
-        # Note: the --no-build-isolation is important! meson-python will
-        # auto-rebuild each time SciPy is imported by the Python interpreter.
-        pip install -e . --no-build-isolation
-
-        # To run the tests for, e.g., the `scipy.linalg` module:
-        pytest scipy/linalg
-
+    Editable installs are supported via ``spin install``.
     When making changes to SciPy code, including to compiled code, there is no
     need to manually rebuild or reinstall. When you run ``git clean -xdf``,
     which removes the built extension modules, remember to also uninstall SciPy
@@ -417,6 +401,12 @@ interface is self-documenting, so please see ``spin --help`` and
 
     See the meson-python_ documentation on editable installs for more details
     on how things work under the hood.
+
+    Note that editable installations are unsuitable for some forms of development,
+    such as working on sections of C/Cython API where tests are disabled for editable
+    installations. They also tend to hit weird corner cases more frequently than
+    regular installations, and have some known limitations like a lack of support
+    for static typing.
 
 
 Installing static type stubs

